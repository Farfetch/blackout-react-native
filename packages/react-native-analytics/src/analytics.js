import AnalyticsCore, {
  trackTypes,
  platformTypes,
} from '@farfetch/blackout-core/analytics';

import * as contexts from './contexts';

class AnalyticsNative extends AnalyticsCore {
  constructor() {
    super(platformTypes.Mobile);

    // Stores the last page call
    this.currentScreenCallData = null;

    // Add default contexts for the mobile platform
    this.useContext(contexts.deviceLanguageContext);
    this.useContext(contexts.deviceModelContext);
    this.useContext(contexts.deviceOSContext);
    this.useContext(contexts.screenDimensionsContext);
  }

  /**
   * Whenever the integrations are loaded at a certain point in time, we fetch them and send the current tracked screen information.
   * This can happen whenever the user gives consent for a specific category mid session.
   *
   * @param {Array<Integration>} loadedIntegrations - An array that contains the integrations that were loaded in runtime, namely after setConsent is called.
   *
   * @async
   * @returns {Promise}                             - Promise that will resolve when the method finishes.
   *
   * @memberof AnalyticsNative#
   */
  async onLoadedIntegrations(loadedIntegrations) {
    // If there is a previous screen call data stored, send a screen event to the integrations that were loaded
    if (this.currentScreenCallData) {
      const { name, properties, eventContext } = this.currentScreenCallData;

      const screenEventData = await super.getTrackEventData(
        trackTypes.SCREEN,
        name,
        properties,
        eventContext,
      );

      super.callIntegrationsMethod(
        loadedIntegrations,
        'track',
        screenEventData,
      );
    }
  }

  /**
   * Track method for custom events.
   *
   * @param {String} event                  - Name of the event.
   * @param {Object} properties             - Properties of the event.
   * @param {Object} eventContext           - Context data that is specific for this event.
   *
   * @async
   * @returns {Promise<AnalyticsNative>}    - Promise that will resolve with the analytics instance.
   *
   * @memberof AnalyticsNative#
   */
  async track(event, properties, eventContext) {
    await super.track(trackTypes.TRACK, event, properties, eventContext);

    return this;
  }

  /**
   * Tracks a screen view and keeps the last call in case
   * there are new integrations that are loaded mid-session
   * ,i.e. the user gives consent.
   *
   * @param {String} name                   - Name of the screen that is to be tracked.
   * @param {Object} properties             - Properties associated with the screen view.
   * @param {Object} eventContext           - Context data that is specific for this event.
   *
   * @async
   * @returns {Promise<AnalyticsNative>}    - Promise that will resolve with the analytics instance.
   *
   * @memberof AnalyticsNative#
   */
  async screen(name, properties, eventContext) {
    // Override the last screen call data with the current one
    this.currentScreenCallData = {
      name,
      properties,
      eventContext,
    };

    await super.track(trackTypes.SCREEN, name, properties, eventContext);
    return this;
  }
}

export default new AnalyticsNative();
