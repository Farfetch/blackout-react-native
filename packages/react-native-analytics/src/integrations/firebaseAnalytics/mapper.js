import eventTypes from '../../eventTypes';
import screenTypes from '../../screenTypes';
import get from 'lodash/get';
import pick from 'lodash/pick';

/**
 * Events mapper that will return a known method from Firebase Analytics package.
 * If Firebase Analytics provides a high-level method like `firebaseAnalytics().logAppOpen()`, then we can map this event with "method" property, along with its properties.
 * If it does not provide a high-level method for a supported event, we can extend this default mapper with the supported event and tell which properties will it send.
 * This event will be sent with `firebaseAnalytics().logEvent(event, properties)`.
 */
export const eventsMapper = {
  // App
  [eventTypes.APP_OPENED]: () => {
    return {
      method: 'logAppOpen',
      properties: undefined,
    };
  },
  // User
  [eventTypes.SIGNUP_FORM_COMPLETED]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logSignUp',
      properties: pick(properties, ['method']),
    };
  },

  // Cart
  [eventTypes.PRODUCT_ADDED_TO_CART]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logAddToCart',
      properties: pick(properties, ['currency', 'items', 'value']),
    };
  },
  [eventTypes.PRODUCT_REMOVED_FROM_CART]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logRemoveFromCart',
      properties: pick(properties, ['currency', 'items', 'value']),
    };
  },
  [screenTypes.BAG]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logViewCart',
      properties: pick(properties, ['currency', 'value', 'items']),
    };
  },

  // Wishlist
  [eventTypes.PRODUCT_ADDED_TO_WISHLIST]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logAddToWishlist',
      properties: pick(properties, ['currency', 'items', 'value']),
    };
  },

  // Product
  [eventTypes.PRODUCT_VIEWED]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logViewItem',
      properties: pick(properties, ['currency', 'value', 'items']),
    };
  },
  [eventTypes.PRODUCT_LIST_VIEWED]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logViewItemList',
      properties: pick(properties, ['item_list_id', 'item_list_name', 'items']),
    };
  },
  [eventTypes.PRODUCT_CLICKED]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logSelectItem',
      properties: pick(properties, [
        'content_type',
        'item_list_id',
        'item_list_name',
        'items',
      ]),
    };
  },

  // Checkout
  [eventTypes.CHECKOUT_STARTED]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logBeginCheckout',
      properties: pick(properties, ['currency', 'value', 'coupon', 'items']),
    };
  },
  [eventTypes.CHECKOUT_STEP_COMPLETED]: data => {
    const properties = get(data, 'properties', {});
    const step = get(properties, 'step', 0);

    switch (Number(step)) {
      case 1:
        return {
          method: 'logAddShippingInfo',
          properties: pick(properties, [
            'currency',
            'value',
            'coupon',
            'shipping_tier',
            'items',
          ]),
        };
      case 2:
        return {
          method: 'logSetCheckoutOption',
          properties: pick(properties, ['checkout_option', 'checkout_step']),
        };
      case 3:
        return {
          method: 'logAddPaymentInfo',
          properties: pick(properties, [
            'currency',
            'value',
            'coupon',
            'payment_type',
            'items',
          ]),
        };
      default:
        return undefined;
    }
  },
  [eventTypes.ORDER_COMPLETED]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logPurchase',
      properties: pick(properties, [
        'transaction_id',
        'affiliation',
        'currency',
        'value',
        'tax',
        'shipping',
        'coupon',
        'items',
      ]),
    };
  },
  [eventTypes.ORDER_REFUNDED]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logRefund',
      properties: pick(properties, [
        'transaction_id',
        'affiliation',
        'currency',
        'value',
        'coupon',
        'shipping',
        'tax',
        'items',
      ]),
    };
  },
  [eventTypes.VIEWED_PROMOTION]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logViewPromotion',
      properties: pick(properties, [
        'creative_name',
        'creative_slot',
        'items',
        'location_id',
        'promotion_id',
        'promotion_name',
      ]),
    };
  },
  [eventTypes.SELECTED_PROMOTION]: data => {
    const properties = get(data, 'properties', {});

    return {
      method: 'logViewPromotion',
      properties: pick(properties, [
        'creative_name',
        'creative_slot',
        'items',
        'location_id',
        'promotion_id',
        'promotion_name',
      ]),
    };
  },
};
