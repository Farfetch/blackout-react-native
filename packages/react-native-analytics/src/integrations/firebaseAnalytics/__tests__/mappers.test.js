import { eventsMapper } from '../mapper';
import { eventTypes } from '@farfetch/blackout-core/analytics';

const mockData = {
  properties: {
    method: 'method',
    item_list_id: 'item_list_id',
    item_list_name: 'item_list_name',
    currency: 'currency',
    value: 'value',
    coupon: 'coupon',
    items: 'items',
    checkout_option: 'checkout_option',
    checkout_step: 'checkout_step',
    creative_name: 'creative_name',
    creative_slot: 'creative_slot',
    location_id: 'location_id',
    promotion_id: 'promotion_id',
    promotion_name: 'promotion_name',
  },
};

describe('eventsMapper', () => {
  Object.keys(eventsMapper).forEach(event => {
    it(`Should match the snapshot for the event: ${event}`, () => {
      expect(eventsMapper[event](mockData)).toMatchSnapshot();
    });
  });

  describe('Should return the right method depending on the payload passed', () => {
    it(`for event: ${eventTypes.CHECKOUT_STEP_COMPLETED} with step "1"`, () => {
      const data = {
        properties: {
          step: 1,
        },
      };

      const firebaseAnalyticsMethod = eventsMapper[
        eventTypes.CHECKOUT_STEP_COMPLETED
      ](data);

      expect(firebaseAnalyticsMethod).toMatchSnapshot();
    });
    it(`for event: ${eventTypes.CHECKOUT_STEP_COMPLETED} with step "2"`, () => {
      const data = {
        properties: {
          step: 2,
        },
      };

      const firebaseAnalyticsMethod = eventsMapper[
        eventTypes.CHECKOUT_STEP_COMPLETED
      ](data);

      expect(firebaseAnalyticsMethod).toMatchSnapshot();
    });
  });

  it(`for event: ${eventTypes.CHECKOUT_STEP_COMPLETED} with step "3"`, () => {
    const data = {
      properties: {
        step: 3,
      },
    };

    const firebaseAnalyticsMethod = eventsMapper[
      eventTypes.CHECKOUT_STEP_COMPLETED
    ](data);

    expect(firebaseAnalyticsMethod).toMatchSnapshot();
  });
});
