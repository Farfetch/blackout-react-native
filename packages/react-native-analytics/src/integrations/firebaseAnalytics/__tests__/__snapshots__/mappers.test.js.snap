// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eventsMapper Should match the snapshot for the event: App Opened 1`] = `
Object {
  "method": "logAppOpen",
  "properties": undefined,
}
`;

exports[`eventsMapper Should match the snapshot for the event: Checkout Started 1`] = `
Object {
  "method": "logBeginCheckout",
  "properties": Object {
    "coupon": "coupon",
    "currency": "currency",
    "items": "items",
    "value": "value",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: Checkout Step Completed 1`] = `undefined`;

exports[`eventsMapper Should match the snapshot for the event: Order Completed 1`] = `
Object {
  "method": "logPurchase",
  "properties": Object {
    "coupon": "coupon",
    "currency": "currency",
    "items": "items",
    "value": "value",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: Order Refunded 1`] = `
Object {
  "method": "logRefund",
  "properties": Object {
    "coupon": "coupon",
    "currency": "currency",
    "items": "items",
    "value": "value",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: Product Added to Cart 1`] = `
Object {
  "method": "logAddToCart",
  "properties": Object {
    "currency": "currency",
    "items": "items",
    "value": "value",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: Product Added to Wishlist 1`] = `
Object {
  "method": "logAddToWishlist",
  "properties": Object {
    "currency": "currency",
    "items": "items",
    "value": "value",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: Product Clicked 1`] = `
Object {
  "method": "logSelectItem",
  "properties": Object {
    "item_list_id": "item_list_id",
    "item_list_name": "item_list_name",
    "items": "items",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: Product List Viewed 1`] = `
Object {
  "method": "logViewItemList",
  "properties": Object {
    "item_list_id": "item_list_id",
    "item_list_name": "item_list_name",
    "items": "items",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: Product Removed from Cart 1`] = `
Object {
  "method": "logRemoveFromCart",
  "properties": Object {
    "currency": "currency",
    "items": "items",
    "value": "value",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: Product Viewed 1`] = `
Object {
  "method": "logViewItem",
  "properties": Object {
    "currency": "currency",
    "items": "items",
    "value": "value",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: Selected Promotion 1`] = `
Object {
  "method": "logViewPromotion",
  "properties": Object {
    "creative_name": "creative_name",
    "creative_slot": "creative_slot",
    "items": "items",
    "location_id": "location_id",
    "promotion_id": "promotion_id",
    "promotion_name": "promotion_name",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: Sign-up Form Completed 1`] = `
Object {
  "method": "logSignUp",
  "properties": Object {
    "method": "method",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: Viewed Promotion 1`] = `
Object {
  "method": "logViewPromotion",
  "properties": Object {
    "creative_name": "creative_name",
    "creative_slot": "creative_slot",
    "items": "items",
    "location_id": "location_id",
    "promotion_id": "promotion_id",
    "promotion_name": "promotion_name",
  },
}
`;

exports[`eventsMapper Should match the snapshot for the event: bag 1`] = `
Object {
  "method": "logViewCart",
  "properties": Object {
    "currency": "currency",
    "items": "items",
    "value": "value",
  },
}
`;

exports[`eventsMapper Should return the right method depending on the payload passed for event: Checkout Step Completed with step "1" 1`] = `
Object {
  "method": "logAddShippingInfo",
  "properties": Object {},
}
`;

exports[`eventsMapper Should return the right method depending on the payload passed for event: Checkout Step Completed with step "2" 1`] = `
Object {
  "method": "logSetCheckoutOption",
  "properties": Object {},
}
`;

exports[`eventsMapper for event: Checkout Step Completed with step "3" 1`] = `
Object {
  "method": "logAddPaymentInfo",
  "properties": Object {},
}
`;
